File: agent_replay/cost.py
"""Token cost calculator for agent-replay.

Uses the pricing manifest to compute costs per step.
Never raises — returns 0.0 for unknown models with a warning.
"""

from __future__ import annotations

import logging

from agent_replay.pricing import load_pricing

logger = logging.getLogger("agent_replay")


def calculate_cost(
    model: str,
    provider: str,
    input_tokens: int,
    output_tokens: int,
) -> float:
    """Calculate the USD cost for a given model call.

    Args:
        model: Model name as returned by the provider (e.g. "claude-sonnet-4-6").
        provider: Provider key (e.g. "anthropic", "openai", "google").
        input_tokens: Number of input tokens consumed.
        output_tokens: Number of output tokens generated.

    Returns:
        Cost in USD. Returns 0.0 if the model is not in the pricing manifest.
    """
    pricing = load_pricing()

    provider_pricing = pricing.get(provider)
    if not provider_pricing:
        logger.warning(
            "agent-replay: unknown provider '%s' — cost will be $0.00",
            provider,
        )
        return 0.0

    model_pricing = provider_pricing.get(model)
    if not model_pricing:
        logger.warning(
            "agent-replay: unknown model '%s/%s' — cost will be $0.00",
            provider,
            model,
        )
        return 0.0

    input_cost = (input_tokens / 1_000_000) * model_pricing.get("input_per_million", 0.0)
    output_cost = (output_tokens / 1_000_000) * model_pricing.get("output_per_million", 0.0)

    return input_cost + output_cost

File: ts/src/cost.ts
/**
 * Token cost calculator for agent-replay TypeScript SDK.
 */

import * as fs from 'fs';
import * as path from 'path';

interface ModelPricing {
  input_per_million: number;
  output_per_million: number;
}

type PricingManifest = Record<string, Record<string, ModelPricing>>;

let cachedPricing: PricingManifest | null = null;

function loadPricing(): PricingManifest {
  if (cachedPricing) return cachedPricing;

  const manifestPath = process.env.AGENT_REPLAY_PRICING
    || path.join(__dirname, '..', '..', 'pricing', 'models.yaml');

  try {
    // Simple YAML parser for our flat structure
    const content = fs.readFileSync(manifestPath, 'utf-8');
    const pricing: PricingManifest = {};
    let currentProvider = '';
    let currentModel = '';

    for (const line of content.split('\n')) {
      const trimmed = line.trimEnd();
      if (trimmed.startsWith('#') || !trimmed) continue;

      const indent = line.length - line.trimStart().length;

      if (indent === 2 && trimmed.endsWith(':') && !trimmed.includes('version')) {
        currentProvider = trimmed.trim().replace(':', '');
        pricing[currentProvider] = {};
      } else if (indent === 4 && trimmed.endsWith(':')) {
        currentModel = trimmed.trim().replace(':', '');
        pricing[currentProvider][currentModel] = { input_per_million: 0, output_per_million: 0 };
      } else if (indent === 6 && trimmed.includes(':')) {
        const [key, value] = trimmed.trim().split(':').map(s => s.trim());
        if (pricing[currentProvider]?.[currentModel]) {
          (pricing[currentProvider][currentModel] as any)[key] = parseFloat(value);
        }
      }
    }

    cachedPricing = pricing;
    return pricing;
  } catch {
    console.warn('agent-replay: failed to load pricing manifest');
    return {};
  }
}

export function calculateCost(
  model: string,
  provider: string,
  inputTokens: number,
  outputTokens: number,
): number {
  const pricing = loadPricing();

  const providerPricing = pricing[provider];
  if (!providerPricing) {
    console.warn(`agent-replay: unknown provider '${provider}'`);
    return 0;
  }

  const modelPricing = providerPricing[model];
  if (!modelPricing) {
    console.warn(`agent-replay: unknown model '${provider}/${model}'`);
    return 0;
  }

  const inputCost = (inputTokens / 1_000_000) * modelPricing.input_per_million;
  const outputCost = (outputTokens / 1_000_000) * modelPricing.output_per_million;

  return inputCost + outputCost;
}
